src/librustc_trans/trans/callee.rs

```diff
@@ -603,11 +603,27 @@ pub fn trans_method_call<'a, 'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
     let _icx = push_ctxt("trans_method_call");
     debug!("trans_method_call(call_expr={})", call_expr.repr(bcx.tcx()));
     let method_call = MethodCall::expr(call_expr.id);
-    let method_ty = (*bcx.tcx().method_map.borrow())[method_call].ty;
+    let (origin, method_ty) =
+        bcx.tcx().method_map
+                 .borrow()
+                 .get(&method_call)
+                 .map(|method| (method.origin.clone(), method.ty))
```

Comment:
Ouch, I have hit this before...
I think I'd prefer it if you matched against `ty::MethodTraitObject(_)` while you have a reference to the method origin, instead of cloning.